### PR TITLE
Feature/object edc

### LINKED ISSUE #34: Implement and test the object pose estimation EKF (Vision-EKF)
**Time estimate:** 20 hours  
**Deadline:**  07.02.22

**Description of task:**
Find and squash the bugs that give the invalid results from the EKF node, and finish implementing the vision-pose estimation filter. As the problem is very likely related to frames being used in the simulator and/or one of the conformed transformation matrices, it is important to test the node using the rosbags taken at the MC-lab.

### LINKED ISSUE #42: Orientation Solution Implementation (Vision-EKF)
Position estimation in the Vision-EKF is fully implemented and working, but the orientation estimates diverges at some points in the runtime. The orientation model Jacobian has been calculated using SimuPy, and is a massive 3x3 matrix, which is incedibly hard to debug.

### PR DESCRIPTION
**Pull request by ivangushkov**
The main development on the vision EKF is done. In connection with the full CV pipeline testing this coming weeks, we should merge this with the main branch. I updated the node and measurement models to better reflect what was achieved on the sim branch, and I streamlined and simplified the code a lot. Would be nice if Krisitan has a look through it, and see if there are any unnecessary definitions (hehe) left in the node. Otherwise Benjamin, please have a look at the structure of the package and say if there is something that needs to change.

#34 
#42

### CODE REVIEW

File: object_edc/ReadMe.txt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Solutions for estimation, detection and classification of objects in the world. Separate but ralated problem to the one of state estimation, so initializing with this a new directory for it. Update this file later.
+
+Pip3/pip requirement:
+pip install dataclasses
```
Comment (theBadMusician) [inline object_edc/ReadMe.txt]: Change the name of the file to "README.md".

File: object_edc/ekf_filtering/ReadMe.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+Simple EKF based solution for filtering CV detections of objects like gate and pole. Includes a very general EKF class that admits different process and measurement models, which is the case here. As of this commit the EKF is technically implemented but it needs integration with ROS.
```
Comment (theBadMusician) [inline object_edc/ekf_filtering/ReadMe.txt]: Change the name of the file to "README.md".

File: object_edc/ekf_filtering/ekf/ekf_python2/analysis_py2.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
+#!/usr/bin/env python
+import numpy as np
+from numpy import ndarray
+import scipy.linalg as la
+from gaussparams_py2 import MultiVarGaussian
+from config import DEBUG
+#from typing import Sequence
```
Comment (theBadMusician) [inline object_edc/ekf_filtering/ekf/ekf_python2/analysis_py2.py]: Remove if not used.

File: object_edc/ekf_filtering/ekf/ekf_python2/analysis_py2.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        np.append(NEES_arr, NEES)
+
+    ANEES = np.average(NEES_arr)
+
+    return ANEES
+
+# def get_RMSE(x_upd_gauss_data: Sequence[MultiVarGaussian],
+#              x _gt_data: Sequence[ndarray]):
+
+#              #TODO
```
Comment (theBadMusician) [inline object_edc/ekf_filtering/ekf/ekf_python2/analysis_py2.py]: Remove if not being worked on for now.

File: object_edc/ekf_filtering/ekf/ekf_python2/dynamicmodels_py2.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+#
+#        # TODO replace this with your own code
+#        first_row = [(Ts**3)/3, 0, (Ts**2)/2, 0]
+#        second_row = [0, (Ts**3)/3, 0, (Ts**2)/2]
+#        third_row = [(Ts**2)/2, 0, Ts, 0]
+#        fourth_row = [0, (Ts**2)/2, 0, Ts]
+#
+#        Q = np.array([first_row, second_row, third_row, fourth_row])*self.sigma_a**2
+#
+#        return Q
```
Comment (theBadMusician) [inline object_edc/ekf_filtering/ekf/ekf_python2/dynamicmodels_py2.py]: Needs to be removed if not used, or uncommented if it's something that's useful to have.

**Comment by ivangushkov** [comment]
Time estimate: 3 hours
ToDo before merge:

- [x] Figure out the orientation error on the gate/plane estimation (look into the quaterninon and its w position, I think the VKF thinks w is at position 4 while pointcloud processing puts it at 1.)
- [x] Change name of package to something that is not object_edc. Suggestions: vkf, object_filtering
- [x] Change src to scripts
- [x] Consider removing unnecessary files and folders (python3, the remains of Sensorfusion course etc.)

**Review by theBadMusician — CHANGES_REQUESTED** [review]
Remove the unused, commented out functions/classes and we're good.

**Review by theBadMusician — APPROVED** [review]
Looking good :+1:  
Ready to merge.

**Pull request merged**